packages/common/core-interfaces/src/erasedType.ts

```diff
@@ -0,0 +1,83 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+/**
+ * "Erased" handle which can be used to expose a opaque/erased version of a type without referencing the actual type.
+ * @remarks
+ * This similar to the standard [type erasure](https://en.wikipedia.org/wiki/Type_erasure) pattern,
+ * but for erasing types at the package boundary.
+ *
+ * Recommended usage is to use `interface` instead of `type` so tooling (such as tsc and refactoring tools)
+ * uses the type name instead of expanding it.
+ *
+ * @example
+ * ```typescript
+ * // public
+ * export interface ErasedMyType extends ErasedType<"myPackage.MyType"> {}
+ * // internal
+ * export interface MyType {
+ * 	example: number;
+ * }
+ * // Usage
+ * function extract(input: ErasedMyType): MyType {
+ * 	return input as unknown as MyType;
+ * }
+ * function erase(input: MyType): ErasedMyType {
+ * 	return input as unknown as ErasedMyType;
+ * }
+ * ```
+ *
+ * See `MakeNominal` for some more details.
+ *
+ * Do not use this class with `instanceof`: this will always be false at runtime,
+ * but the compiler may think it's true in some cases.
+ * @privateRemarks
+ * For this pattern to work well it needs to be difficult for a user of the erased type to
+ * implicitly use something other than a instance received from the package as an instance of the erased type in type safe code.
+ *
+ * This means that this type must not be able to be implicitly converted to from any strong type (not `any` or `never`),
+ * and no amount of auto complete or auto-implement refactoring will produce something that can be used as an erased type.
+ * This is accomplished by:
+ *
+ * 1. requiring that values of this type be an instance of this class.
+ * Typescript does not enforce this requirement for class: only for classes with protected or private members, so such member is included.
+ *
+ * 2. making this class impossible to get an instance of.
+ * This is done by having a private constructor.
+ *
+ * 3. ensuring different erased types also using this library can not be implicitly converted between each-other.
+ * This is done by using the "Name" type parameter.
+ * Note that just having the type parameter is not enough since the presence of type parameters has no impact on implicit conversion in TypeScript:
+ * only the usages of the type parameter matter.
+ *
+ * @sealed
+ * @public
+ */
+export abstract class ErasedType<out Name = unknown> {
```

Comment:
This is the new compile-type typing logic for type erasure. It is derived from a similar library from the tree package which this replaces, but if has been simplified a bit, avoiding dependencies. and has more documentation.